
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Pyicoenrich &mdash; pyicoteo 2.0.5 documentation</title>
    
    <link rel="stylesheet" href="_static/pyramid.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '2.0.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pyicoteo 2.0.5 documentation" href="index.html" />
    <link rel="next" title="Pyicoclip" href="pyicoclip.html" />
    <link rel="prev" title="Pyicoller" href="pyicoller.html" />
<link rel="stylesheet" href="http://fonts.googleapis.com/css?family=Neuton&amp;subset=latin" type="text/css" media="screen" charset="utf-8" />
<link rel="stylesheet" href="http://fonts.googleapis.com/css?family=Nobile:regular,italic,bold,bolditalic&amp;subset=latin" type="text/css" media="screen" charset="utf-8" />
<!--[if lte IE 6]>
<link rel="stylesheet" href="_static/ie6.css" type="text/css" media="screen" charset="utf-8" />
<![endif]-->

  </head>
  <body>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="pyicoclip.html" title="Pyicoclip"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pyicoller.html" title="Pyicoller"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">pyicoteo 2.0.5 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pyicoenrich">
<span id="id1"></span><h1>Pyicoenrich<a class="headerlink" href="#pyicoenrich" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>Enrichment analysis can be applied on any type of -seq data. Pyicoenrich performs enrichment analysis on sequenced reads from two conditions. Like this you can find out how significant the difference of these two conditions is, in terms of the number/density of reads overlapping a region of interest.</p>
<div class="figure">
<img alt="_images/enrichment.png" src="_images/enrichment.png" />
</div>
</div>
<div class="section" id="ma-plot">
<h2>MA Plot<a class="headerlink" href="#ma-plot" title="Permalink to this headline">¶</a></h2>
<p>Pyicoenrich is based on the <a class="reference external" href="http://en.wikipedia.org/wiki/MA_plot">MA-Plot</a>.</p>
</div>
<div class="section" id="region-exploration">
<h2>Region exploration<a class="headerlink" href="#region-exploration" title="Permalink to this headline">¶</a></h2>
<p>If a region file is provided, Pyicoenrich returns for each region a Z-Score (See counts file description) which indicates the enrichment/depletion of condition A over condition B. If no region file is provided, Pyicoenrich provides the options to take the union of reads from both conditions as a region and gives back Z-Scores for the generated regions. As regions with 0 reads in one condition might be especially interesting.</p>
<p>In order to decide what regions are to be explored, you have 3 main options:</p>
<div class="section" id="generate-a-file-with-the-region-magic-flag-and-gtf-file">
<h3>Generate a file with the <tt class="docutils literal"><span class="pre">--region-magic</span></tt> flag and GTF file<a class="headerlink" href="#generate-a-file-with-the-region-magic-flag-and-gtf-file" title="Permalink to this headline">¶</a></h3>
<p>See the <a class="reference internal" href="pyicoregion.html#pyicoregion"><em>Pyicoregion</em></a> documentation for examples on how to use <tt class="docutils literal"><span class="pre">--region-magic</span></tt> flag to automatically explore exons, introns and the whole genome using sliding windows automatically generating your region files from standard GENCODE GTF files.</p>
</div>
<div class="section" id="provide-a-regions-file">
<h3>Provide a regions file<a class="headerlink" href="#provide-a-regions-file" title="Permalink to this headline">¶</a></h3>
<p>If a region file is provided, Pyicoenrich returns for each region a z-Score (among others) which indicates the enrichment/depletion of condition A over condition B. The region file should be in BED format. Also, you may consider only discontinuous regions by using the BED12 format:</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output Pyicoenrich_Kidney_Liver_result_Counts -f bed --region genes.bed</pre>
</div>
</div>
<div class="section" id="do-nothing">
<h3>Do nothing<a class="headerlink" href="#do-nothing" title="Permalink to this headline">¶</a></h3>
<p>Don&#8217;t really know where you want to look yet? If no region file is provided, Pyicoenrich will automatically generate one with taking he union of reads from both conditions as a region and gives back Z-Scores for the generated regions.</p>
<p>The flag <tt class="docutils literal"><span class="pre">--proximity</span></tt> controls the distance with which the regions are considered &#8220;joined&#8221;. Default is 50nt:</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output Pyicoenrich_Kidney_Liver_result -f bed --proximity 50nt</pre>
</div>
<div class="figure">
<img alt="_images/region_definition.png" src="_images/region_definition.png" />
</div>
</div>
</div>
<div class="section" id="pseudocounts-flag">
<h2><tt class="docutils literal"><span class="pre">--pseudocounts</span></tt> flag<a class="headerlink" href="#pseudocounts-flag" title="Permalink to this headline">¶</a></h2>
<p>As regions with 0 reads in one condition might be especially interesting, Pyicoenrich can use pseudocounts, in order to avoid a division by 0: Pyicoenrich calculates the ratio of number of reads in both conditions. As there might not be any reads in a region, Pyicoenrich assumes that there is already 1 read in each region in each condition.</p>
</div>
<div class="section" id="stranded-flag">
<h2><tt class="docutils literal"><span class="pre">--stranded</span></tt> flag<a class="headerlink" href="#stranded-flag" title="Permalink to this headline">¶</a></h2>
<p>To take into consideration reads that coincide with the strand direction of the regions (6th column in your BED6 file)</p>
</div>
<div class="section" id="replica-or-technical-control-swap">
<h2>Replica or technical control (swap)<a class="headerlink" href="#replica-or-technical-control-swap" title="Permalink to this headline">¶</a></h2>
<p>To calculate the Z-Score, Pyicoenrich compares the differences between condition A and condition B with the differences between A and A&#8217; (while A&#8217; is the biological replica of A). If no biological replica is available, Pyicoenrich uses a sample swap as a reference. With sample swap we mean that reads from condition A and B are mixed randomly and divided in two sets (with size of those of A and B). In the two resulting sets we do not expect any significant differences, just like in replicas.</p>
<div class="figure">
<img alt="_images/swap.png" src="_images/swap.png" />
<p class="caption">Technical replica (swap) illustration</p>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<p>With replica:</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output n_norm.enrich -f bed --region genes.bed --replica kidney2.bed</pre>
</div>
<p>Using a swap:</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output n_norm.enrich -f bed --region genes.bed</pre>
</div>
</div>
</div>
<div class="section" id="description-of-the-pyicoenrich-counts-file">
<h2>Description of the pyicoenrich counts file<a class="headerlink" href="#description-of-the-pyicoenrich-counts-file" title="Permalink to this headline">¶</a></h2>
<p>Column description of enrichment result where each line describes a region:</p>
<div class="highlight-python"><pre>TIP: If you want to provide pyicoenrich with your own generated counts file, you only need to provide up to column 6)</pre>
</div>
<ol class="arabic simple">
<li>name                    =  chromosome of region</li>
<li>start                   =  region start</li>
<li>end                     =  region end</li>
<li>name2                   =  alternative label for the region, useful to put the gene name on it</li>
<li>score                   =  Reserved by a &#8221;.&#8221; as it is used by the UCSC browser for coloring.</li>
<li>strand                  =  region strand</li>
<li>signal_a                =  Counts in experiment A (normalized if used)</li>
<li>signal_b                =  Counts in experiment B (normalized if used)</li>
<li>signal_prime_1          =  Counts in experiment A (exactly the same as signal_a) or random background 1 (normalized if used)</li>
<li>signal_prime_2         =  Counts in experiment replica A or random background 2 (normalized if used)</li>
<li>A                      =  (log2(signal_a)+log2(signal_b))/2</li>
<li>M                      =  log2(signal_a/signal_b)</li>
<li>total_reads_a          =  total number of reads in sample a</li>
<li>total_reads_b          =  total number of reads in sample b</li>
<li>num_tags_a             =  number of reads in sample a overlapping the region</li>
<li>num_tags_b             =  number of reads in sample b overlapping the region</li>
<li>A_prime                =  (log2(signal_prime_1)+log2(signal_prime_2))/2</li>
<li>M_prime                =  log2(signal_prime_1/signal_prime_2)</li>
<li>total_reads_a          =  total number of reads in sample a</li>
<li>total_reads_b          =  total number of reads in sample b</li>
<li>total_reads_prime_1    =  total number of reads in sample prime 1</li>
<li>total_reads_prime_2    =  total number of reads in sample prime 2</li>
<li>A_median               =   median of A values in window</li>
<li>mean                       =   mean of M_prime values in window</li>
<li>sd                     =   standard deviation of M_prime values in window</li>
<li>zscore                 =  score for the significance of the difference of enrichment between condition a and b compared to prime 1  and prime 2</li>
</ol>
</div>
<div class="section" id="normalization-methods">
<h2>Normalization methods<a class="headerlink" href="#normalization-methods" title="Permalink to this headline">¶</a></h2>
<p>Pyicoenrich included several popular normalization methods for the counts.</p>
<blockquote>
<div><strong>PUBLIC SERVICE ANNOUNCEMENT:</strong> When dealing with normalization methods, one has to be very careful.
There is no silver bullet, you need to understand your data and then apply the method that is appropriate for it.
If you are in doubt, please consult your local statistician.</div></blockquote>
<div class="section" id="total-reads-normalization-n-norm">
<h3>Total reads normalization (<tt class="docutils literal"><span class="pre">--n-norm</span></tt>)<a class="headerlink" href="#total-reads-normalization-n-norm" title="Permalink to this headline">¶</a></h3>
<p>This normalization will calculate the <em>number of reads per million reads</em> in each region and sample. This is a <em>very simple</em> normalization that tries to correct the bias of comparing different samples by total number of reads. You can activate it with the <tt class="docutils literal"><span class="pre">--n-norm</span></tt> flag.</p>
<p>Example. Using 2 reads files, calculate the enrichment normalizing by N</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output n_norm.enrich -f bed --region genes.bed --n-norm</pre>
</div>
<p>If you want to skip the total reads calculation step, you can provide the total number of reads with the following flags.</p>
<dl class="option">
<dt id="cmdoption--total-reads-a">
<tt class="descname">--total-reads-a</tt><tt class="descclassname"></tt><a class="headerlink" href="#cmdoption--total-reads-a" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption--total-reads-b">
<tt class="descname">--total-reads-b</tt><tt class="descclassname"></tt><a class="headerlink" href="#cmdoption--total-reads-b" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="option">
<dt id="cmdoption--total-reads-replica">
<tt class="descname">--total-reads-replica</tt><tt class="descclassname"></tt><a class="headerlink" href="#cmdoption--total-reads-replica" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>Example:</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output n_norm.enrich -f bed --region genes.bed --n-norm --total-reads-a 120000 --total-reads-b 110000</pre>
</div>
</div>
<div class="section" id="region-length-normalization-len-norm">
<h3>Region length normalization (<tt class="docutils literal"><span class="pre">--len-norm</span></tt>)<a class="headerlink" href="#region-length-normalization-len-norm" title="Permalink to this headline">¶</a></h3>
<p>Calculates the number of reads per <strong>region</strong> kilobase. It aims to correct for regions with different lengths.</p>
<p>NOTE: If possible, try not to mix regions with different lengths.</p>
<p>pyicoenrich -reads kidney1.bed liver1.bed -output n_norm.enrich -f bed &#8211;region genes.bed &#8211;n-norm</p>
</div>
<div class="section" id="rpkm-len-norm-and-n-norm">
<h3>RPKM (<tt class="docutils literal"><span class="pre">--len-norm</span></tt> and <tt class="docutils literal"><span class="pre">--n-norm</span></tt>)<a class="headerlink" href="#rpkm-len-norm-and-n-norm" title="Permalink to this headline">¶</a></h3>
<p>The popular RPKM normalization is the combination of both <tt class="docutils literal"><span class="pre">--n-norm</span></tt> and <tt class="docutils literal"><span class="pre">--len-norm</span></tt>:</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output rpkm_norm.enrich -f bed --region genes.bed --n-norm --len-norm</pre>
</div>
</div>
<div class="section" id="trimmed-means-of-m-values-normalization-tmm-norm">
<h3>Trimmed Means of M values normalization (<tt class="docutils literal"><span class="pre">--tmm-norm</span></tt>)<a class="headerlink" href="#trimmed-means-of-m-values-normalization-tmm-norm" title="Permalink to this headline">¶</a></h3>
<p>As proposed by <a class="reference external" href="http://genomebiology.com/2010/11/3/R25">EdgeR</a>.</p>
<p>This calculates the weighted trimmed mean of the log expression ratios (trimmed mean of M values (TMM)). It is based on the hypothesis that most of your regions do not change, and calculates a normalization factor by excluding the total amount of data.</p>
<p>Important flags.</p>
<dl class="option">
<dt id="cmdoption--a-trim">
<tt class="descname">--a-trim</tt><tt class="descclassname"></tt><a class="headerlink" href="#cmdoption--a-trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Proportion of A values to be discarded when doing the</p>
<p>TMM normalization. [Default 0.05]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption--m-trim">
<tt class="descname">--m-trim</tt><tt class="descclassname"></tt><a class="headerlink" href="#cmdoption--m-trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Proportion of M values to be discarded when doing the</p>
<p>TMM normalization. [Default 0.25]</p>
</dd></dl>

<p>Example: TMM normalization calculated discarding the 20% smaller A (less read coverage) and 5% of the regions with the biggest differences (up and down):</p>
<div class="highlight-python"><pre>pyicoenrich -reads kidney1.bed liver1.bed -output rpkm_norm.enrich -f bed --region genes.bed --tmm-norm --a-trim 0.2 --m-trim 0.05</pre>
</div>
</div>
<div class="section" id="full-quantile-normalization-quant-norm">
<h3>Full quantile normalization (<tt class="docutils literal"><span class="pre">--quant-norm</span></tt>)<a class="headerlink" href="#full-quantile-normalization-quant-norm" title="Permalink to this headline">¶</a></h3>
<p>This method is suitable when your samples have too much variability. As eloquently put by <a class="reference external" href="http://simplystatistics.org/2013/04/26/mindlessly-normalizing-genomics-data-is-bad-but-ignoring-unwanted-variability-can-be-worse/">Simplystatistics</a></p>
</div>
</div>
<div class="section" id="interesting-regions">
<h2><tt class="docutils literal"><span class="pre">--interesting-regions</span></tt><a class="headerlink" href="#interesting-regions" title="Permalink to this headline">¶</a></h2>
<p>Providing a list of interesting regions matching the 4th column of the region or count file will highlight them in the MA plot.</p>
<div class="figure">
<img alt="_images/Enrich_Interesting.png" src="_images/Enrich_Interesting.png" />
<p class="caption">Example of an enrichment output plot using <tt class="docutils literal"><span class="pre">--interesting-regions</span></tt>.</p>
</div>
<p>Example:</p>
<div class="highlight-python"><pre>"""
Region file (regions.bed)
chr1 1 100     region1 0 .
chr1 1000 1100 region2 0 .
chr2 1 100     region3 0 .
...
chrN x y       regionN 0 .

Interesting regions file (interreg.txt)
region4
region10
...
regionZ
"""
pyicoenrich -reads kidney1.bed liver1.bed -output rpkm_norm.enrich -f bed --region genes.bed --interesting-regions interreg.txt</pre>
</div>
</div>
<div class="section" id="credit">
<h2>Credit<a class="headerlink" href="#credit" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Developers: Juan González-Vallinas, Ferran Lloret</li>
<li>Beta Testing: Sonja Althammer, Eneritz Agirre, Nuria Conde Pueyo, Juan González-Vallinas</li>
<li>Benchmarking against other DE methods: Sonja Althammer</li>
<li>Speed and memory performance benchmarking: Juan González-Vallinas</li>
<li>Supervision: Eduardo Eyras</li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Pyicoenrich</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#ma-plot">MA Plot</a></li>
<li><a class="reference internal" href="#region-exploration">Region exploration</a><ul>
<li><a class="reference internal" href="#generate-a-file-with-the-region-magic-flag-and-gtf-file">Generate a file with the <tt class="docutils literal"><span class="pre">--region-magic</span></tt> flag and GTF file</a></li>
<li><a class="reference internal" href="#provide-a-regions-file">Provide a regions file</a></li>
<li><a class="reference internal" href="#do-nothing">Do nothing</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pseudocounts-flag"><tt class="docutils literal"><span class="pre">--pseudocounts</span></tt> flag</a></li>
<li><a class="reference internal" href="#stranded-flag"><tt class="docutils literal"><span class="pre">--stranded</span></tt> flag</a></li>
<li><a class="reference internal" href="#replica-or-technical-control-swap">Replica or technical control (swap)</a><ul>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li><a class="reference internal" href="#description-of-the-pyicoenrich-counts-file">Description of the pyicoenrich counts file</a></li>
<li><a class="reference internal" href="#normalization-methods">Normalization methods</a><ul>
<li><a class="reference internal" href="#total-reads-normalization-n-norm">Total reads normalization (<tt class="docutils literal"><span class="pre">--n-norm</span></tt>)</a></li>
<li><a class="reference internal" href="#region-length-normalization-len-norm">Region length normalization (<tt class="docutils literal"><span class="pre">--len-norm</span></tt>)</a></li>
<li><a class="reference internal" href="#rpkm-len-norm-and-n-norm">RPKM (<tt class="docutils literal"><span class="pre">--len-norm</span></tt> and <tt class="docutils literal"><span class="pre">--n-norm</span></tt>)</a></li>
<li><a class="reference internal" href="#trimmed-means-of-m-values-normalization-tmm-norm">Trimmed Means of M values normalization (<tt class="docutils literal"><span class="pre">--tmm-norm</span></tt>)</a></li>
<li><a class="reference internal" href="#full-quantile-normalization-quant-norm">Full quantile normalization (<tt class="docutils literal"><span class="pre">--quant-norm</span></tt>)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#interesting-regions"><tt class="docutils literal"><span class="pre">--interesting-regions</span></tt></a></li>
<li><a class="reference internal" href="#credit">Credit</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pyicoller.html"
                        title="previous chapter">Pyicoller</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pyicoclip.html"
                        title="next chapter">Pyicoclip</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="pyicoclip.html" title="Pyicoclip"
             >next</a> |</li>
        <li class="right" >
          <a href="pyicoller.html" title="Pyicoller"
             >previous</a> |</li>
        <li><a href="index.html">pyicoteo 2.0.5 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Regulatory Genomics Group.
      Last updated on Jun 21, 2013.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>