
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Pyicos &mdash; pyicoteo 2.0.4 documentation</title>
    
    <link rel="stylesheet" href="_static/pyramid.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '2.0.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pyicoteo 2.0.4 documentation" href="index.html" />
    <link rel="next" title="Pyicotrocol" href="pyicotrocol.html" />
    <link rel="prev" title="Pyicoregion" href="pyicoregion.html" />
<link rel="stylesheet" href="http://fonts.googleapis.com/css?family=Neuton&amp;subset=latin" type="text/css" media="screen" charset="utf-8" />
<link rel="stylesheet" href="http://fonts.googleapis.com/css?family=Nobile:regular,italic,bold,bolditalic&amp;subset=latin" type="text/css" media="screen" charset="utf-8" />
<!--[if lte IE 6]>
<link rel="stylesheet" href="_static/ie6.css" type="text/css" media="screen" charset="utf-8" />
<![endif]-->

  </head>
  <body>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="pyicotrocol.html" title="Pyicotrocol"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pyicoregion.html" title="Pyicoregion"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">pyicoteo 2.0.4 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pyicos">
<span id="id1"></span><h1>Pyicos<a class="headerlink" href="#pyicos" title="Permalink to this headline">¶</a></h1>
<p>Pyicos is a command line utility for the conversion and manipulation of genomic coordinates files. It follows a command/sub-command structure</p>
<p>In the interactive help you can visualize the available commands list:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pyicos</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>If you are interested in the usage of a particular command (for example, &#8216;extend&#8217;) and the meaning of its flags type:</p>
<div class="highlight-python"><pre>pyicos extend -h</pre>
</div>
<p>Here we explain briefly what each subcommand does and we give some examples:</p>
<div class="section" id="convert">
<h2>convert<a class="headerlink" href="#convert" title="Permalink to this headline">¶</a></h2>
<p>Converting a file from one format to another format. Currently supported formats are:</p>
<p>experiment: Bed, Wiggle files (bed_wiggle), SAM, BAM, Eland, bedpk (Pyicos default compressed format), bedspk (Pyicos stranded compressed format)</p>
<p>output: Bed, Wiggle files (bed_wiggle, variable_wiggle), SAM, BAM, Eland, bedpk (Pyicos default compressed format), bedspk (Pyicos stranded compressed format)</p>
<p>This operation is useful if you only want to convert your data to another format. Other operations already include a conversion if you specify different experiment and output formats.</p>
<p>Convert a bed file to a half-open variable wig file:</p>
<div class="highlight-python"><pre>pyicos convert my_experiment.bed my_experiment.wig -f bed -F variable_wig -O</pre>
</div>
<p>Convert all pk files in a folder to bed wig files:</p>
<div class="highlight-python"><pre>pyicos convert my_experiment_folder/ outputfolder/ -f pk -F bed_wig</pre>
</div>
</div>
<div class="section" id="remregions">
<h2>remregions<a class="headerlink" href="#remregions" title="Permalink to this headline">¶</a></h2>
<p>Remove regions that overlap with the regions in the &#8220;black list&#8221; file.</p>
<p>Example:</p>
<div class="highlight-python"><pre>pyicos remregions my_experiment.bed regions.bed my_result.bed --experiment-format bed --open-experiment --region-format bed --open-region --output-format bed --open-output</pre>
</div>
</div>
<div class="section" id="remduplicates">
<h2>remduplicates<a class="headerlink" href="#remduplicates" title="Permalink to this headline">¶</a></h2>
<p>Remove the duplicated reads in a file. A duplicate is a read with the same start position as a read that has already been seen. You can choose how many duplicates you want to tolerate. If you want to keep only one read for a start position, set the duplicates to 0.</p>
<p>Example:</p>
<p>Here we tolerate 1 duplicate so a read can not occur more often than twice:</p>
<div class="highlight-python"><pre>pyicos remduplicates my_experiment.bed my_experiment_1dupl.bed --duplicates 1 -f bed -o -F bed</pre>
</div>
</div>
<div class="section" id="strcorr-strand-correlation">
<span id="strcorr"></span><h2>strcorr (Strand Correlation)<a class="headerlink" href="#strcorr-strand-correlation" title="Permalink to this headline">¶</a></h2>
<p>Finds the optimal extension value by finding the &#8220;gap&#8221; between groups of positive and negative cluster of reads by performing a pearson correlation test.</p>
</div>
<div class="section" id="extend">
<h2>extend<a class="headerlink" href="#extend" title="Permalink to this headline">¶</a></h2>
<p>Extend the reads to the estimated fragment length, taking into consideration if they map to the forward or reverse strand of the reference genome.</p>
<div class="figure">
<img alt="_images/Extend.png" src="_images/Extend.png" />
</div>
<p>Examples:</p>
<p>We have a bed file (half open) with reads between 30 and 50 nucleotides long. We want to extend them all to 150 nucleotides and write the output in bedpk-format to accelerate the successive operations:</p>
<div class="highlight-python"><pre>pyicos extend my_experiment.bed my_experiment_ext.bedpk 150 -f bed -o</pre>
</div>
<p>We do the same with the control:</p>
<div class="highlight-python"><pre>pyicos extend control.bed control_ext.bedpk 150 -f bed -o</pre>
</div>
<p>To visualize the data in a genome browser we set the output to be half-open bed_wig:</p>
<div class="highlight-python"><pre>pyicos extend my_experiment.bed my_experiment_ext.bed_wig 150 -f bed -o -F bed_wig -O</pre>
</div>
</div>
<div class="section" id="subtract">
<h2>subtract<a class="headerlink" href="#subtract" title="Permalink to this headline">¶</a></h2>
<p>Subtract the reads in one file from the reads in another file. Using background data (control) improves the results because the background distribution is not supposed to be normal,
and statistical approaches to obtain this have a limited reach.</p>
<p>The most straightforward approach is to subtract the control from the sample. Make sure the sample has been <strong>normalized</strong> to the control beforehand!
Pyicos maintains a 1bp resolution by subtracting the reads nucleotide by nucleotide, rather than doing a statistical approximation.
Operating with directories will only give appropiate results if the files and the control are paired in alphabetical order.</p>
<p>Example:</p>
<p>Subtract the control from the experiment (both have already been extended, converted to bedpk and normalized):</p>
<div class="highlight-python"><pre>pyicos subtract my_experiment_ext_norm.bedpk control_ext.bedpk my_experiment_ext_norm_subtr.bedpk</pre>
</div>
</div>
<div class="section" id="split">
<h2>split<a class="headerlink" href="#split" title="Permalink to this headline">¶</a></h2>
<p>Split peaks into subpeaks in case they fulfill the criteria.</p>
<p>Criteria: peak has at least two neighboring maxima between which the coverage of reads falls below the threshold. The threshold can be set by the user and it reflects a proportion of the lower maximum.</p>
<p>Output: bedpk or Wiggle files</p>
<div class="figure">
<img alt="_images/Split.png" src="_images/Split.png" />
</div>
<p>Example:</p>
<div class="highlight-python"><pre>pyicos split peaks.bedpk peaks_split.bedpk --split-proportion 0.9</pre>
</div>
</div>
<div class="section" id="discard">
<h2>discard<a class="headerlink" href="#discard" title="Permalink to this headline">¶</a></h2>
<p>Discards peaks that look like artifacts due to the sequencing bias. Here we refer to peaks that look like blocks that result from duplicates (reads with the same start position).</p>
<p>Output: bedpk or Wiggle files</p>
<div class="figure">
<img alt="_images/Artifact.png" src="_images/Artifact.png" />
</div>
<p>Example:</p>
<div class="highlight-python"><pre>pyicos discard peaks.bedpk peaks_discA.bedpk</pre>
</div>
</div>
<div class="section" id="filter">
<h2>filter<a class="headerlink" href="#filter" title="Permalink to this headline">¶</a></h2>
<p>Detect and select significant clusters in the file. There are two steps in this operation: Through the poisson operation the thresholds are determined. Next the peaks are filtered according to these thresholds.</p>
<p>Output: bedpk or Wiggle files</p>
<div class="figure">
<img alt="_images/Filter.png" src="_images/Filter.png" />
</div>
<p>Example:</p>
<div class="highlight-python"><pre>pyicos filter peaks.bedpk significant_peaks.bedpk 150</pre>
</div>
</div>
<div class="section" id="push">
<h2>push<a class="headerlink" href="#push" title="Permalink to this headline">¶</a></h2>
<p>Push the reads in the corresponding strand. If a read doesn&#8217;t have a strand, it will be pushed from left to right.</p>
<p>This operation requires tag-like files (bed, eland, sam).</p>
<p>Example:</p>
<div class="highlight-python"><pre>pyicos push my_experiment.bed my_experiment_pushed100.bed 100 -f bed -F bed</pre>
</div>
</div>
<div class="section" id="credit">
<h2>Credit<a class="headerlink" href="#credit" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Developer: Juan González-Vallinas</li>
<li>Beta Testing: Eneritz Agirre, Sonja Althammer, Juan González-Vallinas</li>
<li>Supervision: Eduardo Eyras</li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Pyicos</a><ul>
<li><a class="reference internal" href="#convert">convert</a></li>
<li><a class="reference internal" href="#remregions">remregions</a></li>
<li><a class="reference internal" href="#remduplicates">remduplicates</a></li>
<li><a class="reference internal" href="#strcorr-strand-correlation">strcorr (Strand Correlation)</a></li>
<li><a class="reference internal" href="#extend">extend</a></li>
<li><a class="reference internal" href="#subtract">subtract</a></li>
<li><a class="reference internal" href="#split">split</a></li>
<li><a class="reference internal" href="#discard">discard</a></li>
<li><a class="reference internal" href="#filter">filter</a></li>
<li><a class="reference internal" href="#push">push</a></li>
<li><a class="reference internal" href="#credit">Credit</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pyicoregion.html"
                        title="previous chapter">Pyicoregion</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pyicotrocol.html"
                        title="next chapter">Pyicotrocol</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="pyicotrocol.html" title="Pyicotrocol"
             >next</a> |</li>
        <li class="right" >
          <a href="pyicoregion.html" title="Pyicoregion"
             >previous</a> |</li>
        <li><a href="index.html">pyicoteo 2.0.4 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Regulatory Genomics Group.
      Last updated on Jun 25, 2013.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>